binding/default_validator.go

```diff
@@ -5,79 +5,169 @@
-	}
+type sliceFieldError struct {
+	validator.FieldError
+	index int
+}
+
+func (fe sliceFieldError) Index() int {
+	return fe.index
+}
+
+func (fe sliceFieldError) Error() string {
+	return fmt.Sprintf("[%d]: %s", fe.index, fe.FieldError.Error())
+}
+
+func (fe sliceFieldError) Unwrap() error {
+	return fe.FieldError
+}
+
+// MapFieldError is returned for invalid map values.
+// It extends validator.FieldError with the key of the failing value.
+type MapFieldError interface {
+	validator.FieldError
+	Key() interface{}
+}
+
+type mapFieldError struct {
+	validator.FieldError
+	key interface{}
+}
+
+func (fe mapFieldError) Key() interface{} {
+	return fe.key
 }
 
-var _ StructValidator = &defaultValidator{}
+func (fe mapFieldError) Error() string {
+	return fmt.Sprintf("[%v]: %s", fe.key, fe.FieldError.Error())
+}
+
+func (fe mapFieldError) Unwrap() error {
+	return fe.FieldError
+}
+
+var _ ContextStructValidator = &defaultValidator{}
 
-// ValidateStruct receives any kind of type, but only performed struct or pointer to struct type.
+// ValidateStruct receives any kind of type, but validates only structs, pointers, slices, arrays, and maps.
 func (v *defaultValidator) ValidateStruct(obj interface{}) error {
+	return v.ValidateStructContext(context.Background(), obj)
+}
+
+func (v *defaultValidator) ValidateStructContext(ctx context.Context, obj interface{}) error {
 	if obj == nil {
 		return nil
 	}
 
 	value := reflect.ValueOf(obj)
 	switch value.Kind() {
 	case reflect.Ptr:
-		return v.ValidateStruct(value.Elem().Interface())
+		return v.ValidateStructContext(ctx, value.Elem().Interface())
 	case reflect.Struct:
-		return v.validateStruct(obj)
+		return v.validateStruct(ctx, obj)
 	case reflect.Slice, reflect.Array:
+		var errs validator.ValidationErrors
+
+		if tag, ok := validatorTags[value.Type()]; ok {
+			if err := v.validateVar(ctx, obj, tag); err != nil {
+				errs = append(errs, err.(validator.ValidationErrors)...) // nolint: errorlint
+			}
+		}
+
 		count := value.Len()
-		validateRet := make(sliceValidateError, 0)
 		for i := 0; i < count; i++ {
-			if err := v.ValidateStruct(value.Index(i).Interface()); err != nil {
-				validateRet = append(validateRet, err)
+			if err := v.ValidateStructContext(ctx, value.Index(i).Interface()); err != nil {
+				for _, fieldError := range err.(validator.ValidationErrors) { // nolint: errorlint
+					errs = append(errs, sliceFieldError{fieldError, i})
+				}
 			}
 		}
-		if len(validateRet) == 0 {
-			return nil
+
+		if len(errs) > 0 {
+			return errs
 		}
-		return validateRet
+		return nil
+	case reflect.Map:
+		var errs validator.ValidationErrors
+
+		if tag, ok := validatorTags[value.Type()]; ok {
+			if err := v.validateVar(ctx, obj, tag); err != nil {
+				errs = append(errs, err.(validator.ValidationErrors)...) // nolint: errorlint
+			}
+		}
+
+		for _, key := range value.MapKeys() {
+			if err := v.ValidateStructContext(ctx, value.MapIndex(key).Interface()); err != nil {
```

Comment:
I think it will be more clear to reverse the condition and add the `continue` statement sin the indentation level be decreased.